src/Approximations/overapproximate.jl

```diff
@@ -333,35 +333,73 @@ function overapproximate(lm::LinearMap{N, <:AbstractZonotope},
+- `X`     -- set
+- `dir`   -- (concrete) direction representation
+- `prune` -- (optional, default: `true`) flag for removing redundant
+             constraints in the end
 
 ### Output
 
 A polyhedron overapproximating the set `X` with the directions from `dir`.
-If the directions are known to be bounded, the result is an `HPolytope`,
-otherwise the result is an `HPolyhedron`.
+The overapproximation is computed using support functions. If the obtained set is
+bounded, the result is an `HPolytope`. Otherwise the result is an `HPolyhedron`.
 """
-function overapproximate(X::LazySet{N}, dir::AbstractDirections{N, VN}) where {N, VN}
-    halfspaces = Vector{LinearConstraint{N, VN}}()
-    sizehint!(halfspaces, length(dir))
-    T = isbounding(dir) ? HPolytope : HPolyhedron
-    H = T(halfspaces)
+function overapproximate(X::LazySet{N}, dir::AbstractDirections{N, VN}; prune::Bool=true) where {N, VN}
+    n = dim(X)
+    H = Vector{LinearConstraint{N, VN}}()
+    sizehint!(H, length(dir))
+
     for d in dir
-        addconstraint!(H, LinearConstraint(d, ρ(d, X)))
+        sf = ρ(d, X)
+        if !isinf(sf)
+            push!(H, LinearConstraint(d, sf))
+        end
+    end
+    if prune
+        remove_redundant_constraints!(H) || throw(ArgumentError("unable to remove redundant constraints"))
+    end
+
+    # if the input is bounded and the directions are bounding => output is bounded
+    # otherwise, check boundedness of the output
+    if (isbounded(X) && isbounding(dir)) || _isbounded_stiemke(HPolyhedron(P))
+        return dim(X) == 2 ? HPolygon(H) : HPolytope(H, check_boundedness=false)
+    else
+        return HPolyhedron(P)
     end
-    return H
 end
 
 # alias with HPolytope type as second argument
-function overapproximate(X::LazySet{N}, ::Type{<:HPolytope}, dirs::AbstractDirections) where {N}
-    return overapproximate(X, dirs)
+function overapproximate(X::LazySet{N}, ::Type{<:HPolytope}, dirs::AbstractDirections{N}; prune::Bool=true) where {N}
+    P = overapproximate(X, dirs, prune=prune)
+    isa(P, HPolytope) || throw(ArgumentError("can't overapproximate with an `HPolytope` " *
+                                "because the set is unbounded; try using an `HPolyhedron`"))
+    return P
+end
+
+# alias with HPolyhedron type as second argument
+function overapproximate(X::LazySet{N}, ::Type{<:HPolyhedron}, dirs::AbstractDirections{N}; prune::Bool=true) where {N}
+    return convert(HPolyhedron, overapproximate(X, dirs, prune=true))
+end
+
+# this function overapproximates a bounded polyhedron with a list of directions
+# that define a bounded set (without checking these assumptions); the result is always bounded
+function _overapproximate_bounded_polyhedron(X::LazySet{N}, dir::AbstractDirections{N, VN}; prune::Bool=true) where {N, VN}
+    H = Vector{LinearConstraint{N, VN}}()
+    sizehint!(H, length(dir))
+    for d in dir
+        sf = ρ(d, X)
+        push!(H, LinearConstraint(d, sf))
```

Comment:
instead of `addconstraint!` (which, for a vector of constraints, only works in 2D) i pushed directly here.